src/mame/taito/qix.h

```diff
@@ -39,140 +34,149 @@ class qix_state : public driver_device
+	uint8_t videoram_r(offs_t offset);
+	void videoram_w(offs_t offset, uint8_t data);
+	uint8_t addresslatch_r();
+	void addresslatch_w(uint8_t data);
+	void paletteram_w(offs_t offset, uint8_t data);
+	void palettebank_w(uint8_t data);
 
 	TIMER_CALLBACK_MEMBER(pia_w_callback);
 	TIMER_CALLBACK_MEMBER(deferred_sndpia1_porta_w);
-	void qix_vsync_changed(int state);
-	void qix_pia_w(offs_t offset, uint8_t data);
-	void qix_coinctl_w(uint8_t data);
+	void vsync_changed(int state);
+	void pia_w(offs_t offset, uint8_t data);
+	void coinctr_w(uint8_t data);
 	void display_enable_changed(int state);
-	void qix_flip_screen_w(int state);
-	void qix_dac_w(uint8_t data);
-	void qix_vol_w(uint8_t data);
+	void flip_screen_w(int state);
+	void dac_w(uint8_t data);
+	void vol_w(uint8_t data);
 	void sndpia_2_warning_w(uint8_t data);
 	void sync_sndpia1_porta_w(uint8_t data);
 	MC6845_BEGIN_UPDATE(crtc_begin_update);
 	MC6845_UPDATE_ROW(crtc_update_row);
-	void set_pen(int offs);
-	int kram3_permut1(int idx, int value);
-	int kram3_permut2(int tbl_index, int idx, const uint8_t *xor_table);
-	int kram3_decrypt(int address, int value);
-	void kram3_lic_maincpu_changed(int state);
-	void kram3_lic_videocpu_changed(int state);
 
 	void audio_map(address_map &map) ATTR_COLD;
-	void kram3_main_map(address_map &map) ATTR_COLD;
-	void kram3_video_map(address_map &map) ATTR_COLD;
-	void main_map(address_map &map) ATTR_COLD;
-	void qix_video_map(address_map &map) ATTR_COLD;
+	virtual void main_map(address_map &map) ATTR_COLD;
+	virtual void video_map(address_map &map) ATTR_COLD;
 };
 
+// with encrypted CPU opcode
+class kram3_state : public qix_state
+{
+public:
+	kram3_state(const machine_config &mconfig, device_type type, const char *tag) :
+		qix_state(mconfig, type, tag),
+		m_mainbank(*this, "mainbank")
+	{ }
+
+	void kram3(machine_config &config);
+	void video(machine_config &config);
+
+	void init_kram3();
+
+protected:
+	virtual void main_map(address_map &map) override ATTR_COLD;
+	virtual void video_map(address_map &map) override ATTR_COLD;
+
+private:
+	void lic_maincpu_changed(int state);
+	void lic_videocpu_changed(int state);
+
+	std::unique_ptr<uint8_t[]> m_main_decrypted;
+	std::unique_ptr<uint8_t[]> m_video_decrypted;
+
+	required_memory_bank m_mainbank;
+};
+
+// with different sound system, video RAM write masking feature
 class slither_state : public qix_state
 {
 public:
 	slither_state(const machine_config &mconfig, device_type type, const char *tag) :
 		qix_state(mconfig, type, tag),
 		m_sn(*this, "sn%u", 1U),
+		m_videoram_mask(*this, "videoram_mask"),
 		m_trak(*this, "AN%u", 0U)
 	{ }
 
 	void slither(machine_config &config);
-	void slither_video(machine_config &config);
-	void slither_audio(machine_config &config);
+	void video(machine_config &config);
+	void audio(machine_config &config);
 
 protected:
 	virtual void machine_start() override ATTR_COLD;
 
+	virtual void video_map(address_map &map) override ATTR_COLD;
```

Comment:
Same here – you’ve got a separate machine configuration function, so you may as well give this a distinct name and set it in machine configuration.